server/trillian_log_server.go

```diff
@@ -56,7 +56,30 @@ func (t *TrillianLogServer) getStorageForLog(logId int64) (storage.LogStorage, e
 
 // QueueLeaves submits a batch of leaves to the log for later integration into the underlying tree.
 func (t *TrillianLogServer) QueueLeaves(ctx context.Context, req *trillian.QueueLeavesRequest) (*trillian.QueueLeavesResponse, error) {
-	return nil, ErrNotImplemented
+	leaves := protosToLeaves(req.Leaves)
+
+	if len(leaves) == 0 {
+		return &trillian.QueueLeavesResponse{Status: buildStatusWithDesc(trillian.TrillianApiStatusCode_ERROR, "Must queue at least one leaf")}, nil
+	}
+
+	tx, err := t.prepareStorageTx(*req.LogId)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = tx.QueueLeaves(leaves)
+
+	if err != nil {
+		tx.Rollback()
```

Comment:
An idea for removing the sprinkling of `Rollback()` calls (and avoid accidentally missing some out) might be to name the return values and then do something like this:
```go
func rollbackTXOnError(tx ..., err error) {
  if err != nil && tx != nil {
    tx.Rollback()
  }
}

func (t *TrillianLogServer) QueueLeaves(...) (resp *trilli..., err error) {
   ...
   tx, err := t.prepareStorageTx(*req.LogId)
   defer rollbackTXOnError(tx, err)
   ...
}
```

Then, if you call return anywhere in the function with a non-nil error, the tx will be rolled-back.